lib/io/spi.cpp

```diff
@@ -0,0 +1,233 @@
+#include "spi.hpp"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+
+namespace hyped::io {
+
+// define what the address space of SPI looks like, see
+// https://github.com/Hyp-ed/hyped-2023/wiki/SPI-Interfacing-on-BBB#spi-register-summary
+#pragma pack(1)
+struct Spi_Channel_Registers {
+  std::uint32_t conf;  // 0x00
+  std::uint32_t stat;  // 0x04
+  std::uint32_t ctrl;  // 0x08
+  std::uint32_t tx;    // 0x0c
+  std::uint32_t rx;    // 0x10
+};
+
+#pragma pack(1)  // ensuring compiler does not add padding
+struct Spi_Registers {
+  std::uint32_t revision;          // 0x000
+  std::uint32_t reserved0[0x43];   // 0x004 - 0x110
+  std::uint32_t sysconfig;         // 0x110
+  std::uint32_t sysstatus;         // 0x114
+  std::uint32_t irqstatus;         // 0x118
+  std::uint32_t irqenable;         // 0x11c
+  std::uint32_t reserved1[2];      // 0x120 - 0x124
+  std::uint32_t syst;              // 0x124
+  std::uint32_t modulctr;          // 0x128
+  Spi_Channel_Registers channel0;  // 0x12c - 0x140
+  Spi_Channel_Registers channel1;  // 0x140 - 0x154
+  Spi_Channel_Registers channel2;  // 0x154 - 0x168
+  Spi_Channel_Registers channel3;  // 0x168 - 0x17c
+  std::uint32_t xferlevel;         // 0x17c
+};
+
+Spi::Spi(core::ILogger &logger)
+    : logger_(logger),
+      file_descriptor_(-1),
+      spi_registers_(0),
+      spi_channel0_registers_(0),
+      has_initialised_(Initialised::kFalse)
+{
+}
+
+Spi::~Spi()
+{
+  close(file_descriptor_);
+}
+
+core::Result Spi::read(std::uint8_t addr, std::uint8_t *rx, std::uint16_t len)
+{
+  if (has_initialised_ == Initialised::kFalse) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read as SPI device has not been initialised");
+    return core::Result::kFailure;
+  }
+  if (file_descriptor_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to open SPI device wile reading");
+    return core::Result::kFailure;
+  }
+  spi_ioc_transfer message[2] = {};
+  // send address
+  message[0].tx_buf = reinterpret_cast<std::uint64_t>(&addr);
+  message[0].rx_buf = 0;
+  message[0].len    = 1;
+  // receive data
+  message[1].tx_buf      = 0;
+  message[1].rx_buf      = reinterpret_cast<std::uint64_t>(rx);
+  message[1].len         = len;
+  const auto read_result = ioctl(file_descriptor_, SPI_IOC_MESSAGE(2), message);
+  if (read_result < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read from SPI device");
+    return core::Result::kFailure;
+  }
+  logger_.log(core::LogLevel::kDebug, "Successfully read from SPI device");
+  return core::Result::kSuccess;
+}
+
+core::Result Spi::write(std::uint8_t addr, std::uint8_t *tx, std::uint16_t len)
+{
+  if (has_initialised_ == Initialised::kFalse) {
```

Comment:
this is why i don't like this flag-style approach; once the caller is in possession of an `Spi` object it should just work.

Reply:
yeah will make the static fucntion to return `Spi` as well